### PR TITLE
EAR 264 - Validate when answer moved/deleted referenced by user validations

### PR DESCRIPTION
**Pull request by tbaggaley**
## What is the context of this PR?

Adds validation errors to the "PreviousAnswer" option available when setting user validations for all applicable types: Date answers, Numeric answers and pages' Total validations.

Refactors the AJV schema to facilitate this - `entities.json` has been broken down into several smaller sub-schemas which reference each other for ease of management / editing; the duplication from answer validations in the schema has been removed where possible so that most of the user validation validating logic occurs within one definition.

Some minor style changes: total validation modal width increased slightly so that the error message fits when previous answers are moved.

### API

- [x] Add generic custom keywords (idExists & idPreceedsEntity) w/ appropriate data structure (memoized)
- [x] Add AJV validation for totalValidation - answer deleted/moved
- [x] Add tests for totalValidation validation
- [x] Split up existing AJV schema for ease of handling
- [x] Factor out common validation code from answer types
- [x] Add AJV validation for numericAnswer, dateAnswer and dateRange answer validation
- [x] Add tests for validation re: moving / deleting previous answers

#### Numeric Validation

-   [x] NumericValidation: show errors on PreviousAnswer box when not selected (`ERR_NO_VALUE`) - currently has no error styling / message.
-   [x] Show errors on PreviousAnswer box when `ERR_REFERENCE_DELETED` or `ERR_REFERENCE_MOVED`
-   [x] Fix validation modal not opening on correct tab (should match button pushed)

<a id="org7c5fbf3"></a>

#### Date Validation

-   [x] Show errors on PreviousAnswer box when not selected (`ERR_NO_VALUE`) - currently has no error styling / message
-   [x] Show errors on PreviousAnswer box when `ERR_REFERENCE_DELETED` or `ERR_REFERENCE_MOVED`
-   [x] Show errors on &ldquo;offset&rdquo; value box *iff* `ERR_OFFSET_NO_VALUE` present

<a id="org4d2ca82"></a>

#### Total Validation

-   [x] Show errors on PreviousAnswer box when not selected (`ERR_NO_VALUE`) - currently has no error styling / message
-   [x] Show errors on PreviousAnswer box when `ERR_REFERENCE_DELETED` or `ERR_REFERENCE_MOVED`

## Refs

- [Ticket](https://collaborate2.ons.gov.uk/jira/browse/EAR-264)
- [Design](https://overflow.io/s/0IT5FFVP?node=421703b4&prototype)

## How to review
- Open your favourite questionnaire
- Add a validation rule to any numeric answer (`Currency`, `Percentage`, `Number`, `Unit`, etc.)
- Should be an error re: no value / answer selected, both on the question page and in the modal for the validation
- Select a previous answer
- Error goes away
- Delete the question referred to
- Error appears both on question page & validation modal stating that referenced answer has been deleted
- Select another previous answer
- Move the previous answer to after the question using it as a reference
- Error appears on both question page & validation modal stating that referenced answer has been moved
- Rinse & repeat the above with `Date` answers
- Rinse & repeat the above with `Total` validation

### What to do after everything is green

1. - Bring this branch up-to-date with Origin/Master
2. - If there are a lot of commits or some are not helpful to read, squash them down
3. - Click the **Merge** button on this pull request
4. - Does this change mean the **Capability examples** questionnaire should be updated?
5. - Move the Jira ticket for this task into the next stage of the process

### CODE REVIEW

File: eq-author/src/App/page/Design/Validation/PreviousAnswerEditor.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  ERR_REFERENCE_MOVED,
+};
+
+const PreviousAnswerEditor = ({
+  onChangeUpdate,
+  answer,
+  validation,
+  readKey,
+}) => {
+  const { errors } = validation.validationErrorInfo;
```
Comment (tjamesmac) [inline eq-author/src/App/page/Design/Validation/PreviousAnswerEditor.js:26]:
```suggestion
  const { errors } = validation?.validationErrorInfo;
```

Comment (tjamesmac) [reply eq-author/src/App/page/Design/Validation/PreviousAnswerEditor.js:26]: ?

File: eq-author-api/src/validation/customKeywords/idPreceedsCurrentEntity.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      dataPath,
+      parentData,
+      fieldName,
+      questionnaire
+    ) => {
+      const ourId = currentSchema.id || parentData.id;
+      const ourPosition = getAbsolutePositionById({ questionnaire }, ourId);
+      const theirPosition = getAbsolutePositionById({ questionnaire }, theirId);
+
+      // This keyword passes if target ID doesn't exist in order to simply existing AJV schema
```
Comment (tjamesmac) [inline eq-author-api/src/validation/customKeywords/idPreceedsCurrentEntity.js:19]:
```suggestion
      // This keyword passes if target ID doesn't exist in order to simplify existing AJV schema
```

Comment (tjamesmac) [reply eq-author-api/src/validation/customKeywords/idPreceedsCurrentEntity.js:19]: ?

**Review by zimboflyman — DISMISSED** [review]
Looks good - is it me, or does the validation seem quicker now? Just seems smoother - nice

**Comment by tbaggaley** [comment]
> Looks good - is it me, or does the validation seem quicker now? Just seems smoother - nice

Yeah, I think it is - not sure why exactly, but after splitting up the schemas etc. the validation tests run much faster too. Used to take 30s to run the validation tests file - now takes less than 2s

File: eq-author/src/App/page/Design/Validation/DateValidation/index.js [hunk 3]
```diff
@@ -65,15 +79,23 @@ const UnwrappedDateValidation = ({ validation, answer, displayName, onChange, on
...
+            <PositionPicker
+              value={validation.relativePosition}
+              onChange={onChange}
+              onUpdate={onUpdate}
+            />
+          ) : (
+            <RelativePositionText>
+              {" "}
+              {validation.relativePosition.toLowerCase()}{" "}
+            </RelativePositionText>
```
Comment (tjamesmac) [inline eq-author/src/App/page/Design/Validation/DateValidation/index.js:97]: Are the `{" "}` for spaces?

Comment (tbaggaley) [reply eq-author/src/App/page/Design/Validation/DateValidation/index.js:97]: No, the linter inserts them for some reason - I keep removing them and they go back :-(

File: eq-author/src/App/page/Design/Validation/DateValidation/index.js [hunk 4]
```diff
@@ -65,15 +79,23 @@ const UnwrappedDateValidation = ({ validation, answer, displayName, onChange, on
...
-          }
+          {answer.type === DATE ? (
+            <PositionPicker
+              value={validation.relativePosition}
+              onChange={onChange}
+              onUpdate={onUpdate}
+            />
+          ) : (
+            <RelativePositionText>
+              {" "}
```
Comment (tbaggaley) [inline eq-author/src/App/page/Design/Validation/DateValidation/index.js:95]:
```suggestion
```

File: eq-author/src/App/page/Design/Validation/DateValidation/index.js [hunk 5]
```diff
@@ -65,15 +79,23 @@ const UnwrappedDateValidation = ({ validation, answer, displayName, onChange, on
...
+              {validation.relativePosition.toLowerCase()}{" "}
```
Comment (tbaggaley) [inline eq-author/src/App/page/Design/Validation/DateValidation/index.js:96]:
```suggestion
              {validation.relativePosition.toLowerCase()}
```

File: eq-author/src/App/page/Design/Validation/DateValidation/index.js (hunk 3 above)
Comment (tbaggaley) [reply eq-author/src/App/page/Design/Validation/DateValidation/index.js]: There we go - hopefully they're gone for good now

**Review by zimboflyman — APPROVED** [review]

**Pull request merged**